In [1]:
# Importing Required Libraries
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np
import requests

In [2]:
file = pd.read_csv("internal_link_file.csv")

In [3]:
file.shape

(2609, 2)

##### Here I am scrapping all billionaires personal data and their wealth history using "internal_link_file.csv" that contains about 2609 links where all billionaires personal data with their wealth history are available.

##### But before scraping all the data, lets first scrape as test of 1st billionaire data.

In [4]:
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36',"Content-Type": "application/x-www-form-urlencoded"} 
url = "https://www.forbes.com/profile/elon-musk/?list=rtb/"
webpage = requests.get(url, headers=headers).text

In [5]:
soup = BeautifulSoup(webpage, 'lxml')

In [6]:
print(soup.prettify())

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <meta content="en_US" http-equiv="Content-Language"/>
  <meta content="width=device-width,initial-scale=1,maximum-scale=5,minimum-scale=1,user-scalable=yes" name="viewport"/>
  <title itemprop="headline">
   Pham Nhat Vuong
  </title>
  <link href="https://i.forbesimg.com/48X48-F.png" rel="shortcut icon"/>
  <link rel="tttt"/>
  <link href="https://www.forbes.com/profile/pham-nhat-vuong/" rel="canonical"/>
  <meta content="origin-when-cross-origin" name="referrer"/>
  <meta content="#699 Pham Nhat Vuong on the Real-Time Billionaires - Pham Nhat Vuong studied in Russia and started a popular instant noodle business in the Ukraine in" itemprop="description" name="description"/>
  <meta content="Pham Nhat Vuong" name="keywords"/>
  <meta content="Forbes" property="og:site_name"/>
  <meta content="Pham Nhat Vuong" property="og:title"/>
  <meta content="website" property="og:type"/>
  <meta content="https://www.forbes.com/p

In [7]:
# name
soup.find('h1','listuser-header__name').text

'Pham Nhat Vuong'

In [135]:
# Personal Stats
soup.find('dl','listuser-block').text

'Age52Source of WealthTesla, SpaceX, Self MadeSelf-Made Score8ResidenceAustin, TexasCitizenshipUnited StatesMarital StatusSingleChildren10EducationBachelor of Arts/Science, University of Pennsylvania'

In [170]:
# Wealth History
soup.find('canvas').get('data-chart')

'[{"percent":13.836477987421384,"worth":"$2.2B","name":"Billionaires","date":"March 2018"},{"percent":20.125786163522015,"worth":"$3.2B","name":"Billionaires","date":"March 2019"},{"percent":15.09433962264151,"worth":"$2.4B","name":"Billionaires","date":"April 2020"},{"percent":100,"worth":"$15.9B","name":"Billionaires","date":"April 2021"},{"percent":54.088050314465406,"worth":"$8.6B","name":"Billionaires","date":"April 2022"},{"percent":18.867924528301888,"worth":"$3B","name":"Billionaires","date":"April 2023"}]'

# Main File -  Final Scraping

In [45]:
name = []
wealth_history = []
personal_stats = []
x = file['internal_link'][1189:]
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36',"Content-Type": "application/x-www-form-urlencoded"} 
for i in x:
    print(i)
    url = i
    webpage = requests.get(url, headers=headers).text # fetching html webpage
    soup = BeautifulSoup(webpage, 'lxml')  # giving webpage to soup
    
    try:
        name.append(soup.find('h1','listuser-header__name').text) # name
    except:
        name.append(np.nan)
        
    try:
        wealth_history.append(soup.find('canvas').get('data-chart')) # wealth_history
    except:
        wealth_history.append(np.nan)
        
    try:
        personal_stats.append(soup.find('dl','listuser-block').text) # personal_stats
    except:
        personal_stats.append(np.nan)
        
print("Done")        

https://www.forbes.com/profile/c-dean-metropoulos/?list=rtb/
https://www.forbes.com/profile/lachhman-das-mittal/?list=rtb/
https://www.forbes.com/profile/paolo-rocca/?list=rtb/
https://www.forbes.com/profile/gianfelice-rocca-1/?list=rtb/
https://www.forbes.com/profile/fernando-chico-pardo/?list=rtb/
https://www.forbes.com/profile/liu-xiaodong/?list=rtb/
https://www.forbes.com/profile/ray-davis/?list=rtb/
https://www.forbes.com/profile/tench-coxe/?list=rtb/
https://www.forbes.com/profile/zhu-xingming/?list=rtb/
https://www.forbes.com/profile/alberto-siccardi/?list=rtb/
https://www.forbes.com/profile/prathap-reddy/?list=rtb/
https://www.forbes.com/profile/michael-s-smith/?list=rtb/
https://www.forbes.com/profile/henry-laufer/?list=rtb/
https://www.forbes.com/profile/chen-shibin-1/?list=rtb/
https://www.forbes.com/profile/felix-baker/?list=rtb/
https://www.forbes.com/profile/alexander-ponomarenko/?list=rtb/
https://www.forbes.com/profile/bruce-cheng/?list=rtb/
https://www.forbes.com/profi

https://www.forbes.com/profile/zhao-lixin/?list=rtb/
https://www.forbes.com/profile/david-nahmad/?list=rtb/
https://www.forbes.com/profile/mustafa-rahmi-koc/?list=rtb/
https://www.forbes.com/profile/bruce-karsh/?list=rtb/
https://www.forbes.com/profile/tor-peterson/?list=rtb/
https://www.forbes.com/profile/eduardo-eurnekian/?list=rtb/
https://www.forbes.com/profile/tang-yiu/?list=rtb/
https://www.forbes.com/profile/georg-haub/?list=rtb/
https://www.forbes.com/profile/arvind-tiku/?list=rtb/
https://www.forbes.com/profile/filiz-sahenk/?list=rtb/
https://www.forbes.com/profile/william-boyd/?list=rtb/
https://www.forbes.com/profile/william-stone/?list=rtb/
https://www.forbes.com/profile/li-xiaohua/?list=rtb/
https://www.forbes.com/profile/georg-von-opel/?list=rtb/
https://www.forbes.com/profile/robert-duggan/?list=rtb/
https://www.forbes.com/profile/ty-tsai/?list=rtb/
https://www.forbes.com/profile/domenico-dolce/?list=rtb/
https://www.forbes.com/profile/stefano-gabbana/?list=rtb/
https://

https://www.forbes.com/profile/douglas-hsu/?list=rtb/
https://www.forbes.com/profile/anu-aga/?list=rtb/
https://www.forbes.com/profile/alexander-mamut/?list=rtb/
https://www.forbes.com/profile/thomas-kwok/?list=rtb/
https://www.forbes.com/profile/zygmunt-solorz-zak/?list=rtb/
https://www.forbes.com/profile/thaksin-shinawatra/?list=rtb/
https://www.forbes.com/profile/richard-saghian/?list=rtb/
https://www.forbes.com/profile/ion-tiriac/?list=rtb/
https://www.forbes.com/profile/andrey-andreev/?list=rtb/
https://www.forbes.com/profile/drew-houston/?list=rtb/
https://www.forbes.com/profile/david-adelman/?list=rtb/
https://www.forbes.com/profile/mohamed-al-fayed/?list=rtb/
https://www.forbes.com/profile/michael-ashcroft/?list=rtb/
https://www.forbes.com/profile/alex-beard/?list=rtb/
https://www.forbes.com/profile/heikki-kyostila/?list=rtb/
https://www.forbes.com/profile/wu-zhigang/?list=rtb/
https://www.forbes.com/profile/martin-lau/?list=rtb/
https://www.forbes.com/profile/pp-reddy/?list=rt

https://www.forbes.com/profile/lu-yiwen/?list=rtb/
https://www.forbes.com/profile/roberto-hernandez-ramirez/?list=rtb/
https://www.forbes.com/profile/chen-qiongxiang/?list=rtb/
https://www.forbes.com/profile/amy-wyss/?list=rtb/
https://www.forbes.com/profile/hamdi-ulukaya/?list=rtb/
https://www.forbes.com/profile/karl-johan-persson/?list=rtb/
https://www.forbes.com/profile/martin-viessmann/?list=rtb/
https://www.forbes.com/profile/palmer-luckey/?list=rtb/
https://www.forbes.com/profile/timothy-springer/?list=rtb/
https://www.forbes.com/profile/charlotte-colket-weber/?list=rtb/
https://www.forbes.com/profile/jose-roberto-marinho/?list=rtb/
https://www.forbes.com/profile/joao-roberto-marinho/?list=rtb/
https://www.forbes.com/profile/hao-hong/?list=rtb/
https://www.forbes.com/profile/joe-lacob/?list=rtb/
https://www.forbes.com/profile/robert-g-miller/?list=rtb/
https://www.forbes.com/profile/todd-wagner/?list=rtb/
https://www.forbes.com/profile/anatoly-sedykh/?list=rtb/
https://www.forbes

https://www.forbes.com/profile/gerry-harvey/?list=rtb/
https://www.forbes.com/profile/abhay-soi/?list=rtb/
https://www.forbes.com/profile/zbigniew-juroszek/?list=rtb/
https://www.forbes.com/profile/duncan-macmillan/?list=rtb/
https://www.forbes.com/profile/kc-liu/?list=rtb/
https://www.forbes.com/profile/chuanhua-xu/?list=rtb/
https://www.forbes.com/profile/david-booth/?list=rtb/
https://www.forbes.com/profile/chris-wallin/?list=rtb/
https://www.forbes.com/profile/shi-yifeng/?list=rtb/
https://www.forbes.com/profile/eiichi-kuriwada/?list=rtb/
https://www.forbes.com/profile/hua-shen/?list=rtb/
https://www.forbes.com/profile/massimo-moratti/?list=rtb/
https://www.forbes.com/profile/ben-silbermann/?list=rtb/
https://www.forbes.com/profile/stefan-olsson/?list=rtb/
https://www.forbes.com/profile/mario-gabelli/?list=rtb/
https://www.forbes.com/profile/xinghai-zhang/?list=rtb/
https://www.forbes.com/profile/huwen-chen/?list=rtb/
https://www.forbes.com/profile/basudeo-singh/?list=rtb/
https://

https://www.forbes.com/profile/stephen-jarislowsky/?list=rtb/
https://www.forbes.com/profile/tianjiang-jia/?list=rtb/
https://www.forbes.com/profile/sebastian-kulczyk/?list=rtb/
https://www.forbes.com/profile/dan-wilks/?list=rtb/
https://www.forbes.com/profile/katharina-andresen/?list=rtb/
https://www.forbes.com/profile/alexandra-andresen/?list=rtb/
https://www.forbes.com/profile/peter-jackson/?list=rtb/
https://www.forbes.com/profile/william-koch/?list=rtb/
https://www.forbes.com/profile/huxiong-chen/?list=rtb/
https://www.forbes.com/profile/vincent-lo/?list=rtb/
https://www.forbes.com/profile/yuri-shefler/?list=rtb/
https://www.forbes.com/profile/mohammed-dewji/?list=rtb/
https://www.forbes.com/profile/satish-reddy/?list=rtb/
https://www.forbes.com/profile/david-penaloza-alanis/?list=rtb/
https://www.forbes.com/profile/patricia-angelini-rossi/?list=rtb/
https://www.forbes.com/profile/terry-snow/?list=rtb/
https://www.forbes.com/profile/stefan-pierer/?list=rtb/
https://www.forbes.com/

https://www.forbes.com/profile/wu-li-gann/?list=rtb/
https://www.forbes.com/profile/alberto-cortina/?list=rtb/
https://www.forbes.com/profile/ayman-hariri/?list=rtb/
https://www.forbes.com/profile/ali-ghodsi/?list=rtb/
https://www.forbes.com/profile/guo-zhenyu/?list=rtb/
https://www.forbes.com/profile/jinsheng-ren/?list=rtb/
https://www.forbes.com/profile/ashwin-desai/?list=rtb/
https://www.forbes.com/profile/gao-yi/?list=rtb/
https://www.forbes.com/profile/wu-xushun/?list=rtb/
https://www.forbes.com/profile/richard-hayne/?list=rtb/
https://www.forbes.com/profile/richard-chang/?list=rtb/
https://www.forbes.com/profile/li-qiong/?list=rtb/
https://www.forbes.com/profile/anatoly-lomakin/?list=rtb/
https://www.forbes.com/profile/jin-lei/?list=rtb/
https://www.forbes.com/profile/jianyi-wang/?list=rtb/
https://www.forbes.com/profile/itamar-locks/?list=rtb/
https://www.forbes.com/profile/blairo-maggi/?list=rtb/
https://www.forbes.com/profile/g-gnanalingam/?list=rtb/
https://www.forbes.com/pro

https://www.forbes.com/profile/hal-steinbrenner/?list=rtb/
https://www.forbes.com/profile/jessica-steinbrenner/?list=rtb/
https://www.forbes.com/profile/jennifer-steinbrenner-swindal/?list=rtb/
https://www.forbes.com/profile/he-zhaoxi/?list=rtb/
https://www.forbes.com/profile/robert-toennies/?list=rtb/
https://www.forbes.com/profile/clemens-toennies/?list=rtb/
https://www.forbes.com/profile/seth-klarman/?list=rtb/
https://www.forbes.com/profile/tingdong-yang/?list=rtb/
https://www.forbes.com/profile/rajesh-sharma/?list=rtb/
https://www.forbes.com/profile/martin-moller-nielsen/?list=rtb/
https://www.forbes.com/profile/chua-thian-poh/?list=rtb/
https://www.forbes.com/profile/carsten-maschmeyer/?list=rtb/
https://www.forbes.com/profile/r-marcelo-claure/?list=rtb/
https://www.forbes.com/profile/annalisa-doris/?list=rtb/
https://www.forbes.com/profile/massimo-doris/?list=rtb/
https://www.forbes.com/profile/henry-engelhardt/?list=rtb/
https://www.forbes.com/profile/shao-jianxiong/?list=rtb/


https://www.forbes.com/profile/daniel-feffer/?list=rtb/
https://www.forbes.com/profile/ye-fan/?list=rtb/
https://www.forbes.com/profile/ruben-vardanyan/?list=rtb/
https://www.forbes.com/profile/li-yongqing/?list=rtb/
https://www.forbes.com/profile/kenzo-tsujimoto/?list=rtb/
https://www.forbes.com/profile/thongma-vijitpongpun/?list=rtb/
https://www.forbes.com/profile/ruben-feffer/?list=rtb/
https://www.forbes.com/profile/christian-herz/?list=rtb/
https://www.forbes.com/profile/michaela-herz/?list=rtb/
https://www.forbes.com/profile/anant-asavabhokhin-1/?list=rtb/
https://www.forbes.com/profile/giuliana-caprotti/?list=rtb/
https://www.forbes.com/profile/marina-caprotti/?list=rtb/
https://www.forbes.com/profile/gabriella-meister/?list=rtb/
https://www.forbes.com/profile/danna-azrieli/?list=rtb/
https://www.forbes.com/profile/sharon-azrieli/?list=rtb/
https://www.forbes.com/profile/naomi-azrieli/?list=rtb/
https://www.forbes.com/profile/thomas-bailey/?list=rtb/
https://www.forbes.com/profi

https://www.forbes.com/profile/xu-shijun/?list=rtb/
https://www.forbes.com/profile/luigi-cremonini/?list=rtb/
https://www.forbes.com/profile/wanmao-xu/?list=rtb/
https://www.forbes.com/profile/bob-muglia/?list=rtb/
https://www.forbes.com/profile/peter-kamprad/?list=rtb/
https://www.forbes.com/profile/anna-maria-braun/?list=rtb/
https://www.forbes.com/profile/johanna-braun/?list=rtb/
https://www.forbes.com/profile/karl-friedrich-braun/?list=rtb/
https://www.forbes.com/profile/zhang-tianyu/?list=rtb/
https://www.forbes.com/profile/james-truchard/?list=rtb/
https://www.forbes.com/profile/analjit-singh/?list=rtb/
https://www.forbes.com/profile/oleg-boyko/?list=rtb/
https://www.forbes.com/profile/pavel-golubkov/?list=rtb/
https://www.forbes.com/profile/tiger-woods/?list=rtb/
https://www.forbes.com/profile/said-gutseriev/?list=rtb/
https://www.forbes.com/profile/ma-xiuhui/?list=rtb/
https://www.forbes.com/profile/koo-bon-sik/?list=rtb/
https://www.forbes.com/profile/chen-xuehua/?list=rtb/
ht

In [48]:
internal_file_data = pd.DataFrame({
    "name":name,
    "Wealth_History":wealth_history,
    "personal_data":personal_stats
})

In [57]:
internal_file_data

name  \
0                    Elon Musk   
1     Bernard Arnault & family   
2                   Jeff Bezos   
3                Larry Ellison   
4                   Bill Gates   
...                        ...   
2604                Chang Jing   
2605      Vijay Shekhar Sharma   
2606                Kanye West   
2607         Sam Bankman-Fried   
2608          Elizabeth Holmes   

                                         Wealth_History  \
0     [{"percent":3.8356164383561646,"worth":"$8.4B"...   
1     [{"percent":15.876777251184834,"worth":"$33.5B...   
2     [{"percent":18.07909604519774,"worth":"$32B","...   
3     [{"percent":44.85981308411215,"worth":"$48B","...   
4     [{"percent":58.91472868217055,"worth":"$76B","...   
...                                                 ...   
2604  [{"percent":100,"worth":"$2.5B","name":"Billio...   
2605  [{"percent":50,"worth":"$1.3B","name":"Billion...   
2606  [{"percent":90,"worth":"$1.8B","name":"Billion...   
2607  [{"percent":36.25,"worth":"$8.7B","name":"Bill...   
2608  [{"percent":100,"worth":"$4.5B","name":"Billio...   

                                          personal_data  
0     Age52Source of WealthTesla, SpaceX, Self MadeS...  
1     Age74Source of WealthLVMHResidenceParis, Franc...  
2     Age59Source of WealthAmazon, Self MadeSelf-Mad...  
3     Age78Source of WealthOracle, Self MadeSelf-Mad...  
4     Age67Source of WealthMicrosoft, Self MadeSelf-...  
...                                                 ...  
2604  Age40Source of WealthTechnology, Self MadeResi...  
2605  Age45Source of Wealthfinancial technology, Sel...  
2606  Age46Source of Wealthmusic, sneakers, Self Mad...  
2607  Age31Source of Wealthcryptocurrency exchange, ...  
2608  Age39Source of Wealthblood testing, Self MadeS...  

[2609 rows x 3 columns]

In [58]:
internal_file_data.to_csv('internal_file_data.csv', encoding='utf-8')

## Thank You